libogc/sha.c

```diff
@@ -132,29 +155,37 @@ s32 SHA_InitializeContext(sha_context* context)
 	iosFree(__sha_hid, params);
 	return ret;
 }
-s32 SHA_Calculate(sha_context* context, const void* data, const u32 data_size, void* message_digest)
+
+s32 SHA_Input(sha_context* context, const void* data, const u32 data_size)
 {
-	if(context == NULL || message_digest == NULL || data_size == 0 || data == NULL)
+	if(context == NULL || data == NULL || data_size == 0)
 		return -1;
 	
-	if(((u32)context & 0x1F) != 0 || ((u32)message_digest & 0x1F) != 0)
-		return -4;
-	
-	if( data != NULL && ((u32)data & 0x3F) != 0)
+	if(((u32)context) & 0x1F)
 		return -4;
 	
-	return SHA_ExecuteCommand(FinalizeHash, context, data, data_size, message_digest);
+	return SHA_ExecuteCommand(AddData, context, data, data_size, NULL);
 }
 
-s32 SHA_Input(sha_context* context, const void* data, const u32 data_size)
+s32 SHA_Finalize(sha_context* context, const void* data, const u32 data_size, void* message_digest)
 {
-	if(context == NULL || data == NULL || data_size == 0)
+	if(context == NULL || message_digest == NULL || data_size == 0 || data == NULL)
 		return -1;
 	
-	if((((u32)context) & 0x1F) || (((u32)data) & 0x3F))
+	if(((u32)context & 0x1F) != 0 || ((u32)message_digest & 0x1F) != 0)
 		return -4;
 	
-	return SHA_ExecuteCommand(AddData, context, data, data_size, NULL);
+	return SHA_ExecuteCommand(FinalizeHash, context, data, data_size, message_digest);
+}
+
+s32 SHA_Calculate(const void* data, const u32 data_size, void* message_digest)
+{
+	sha_context context ATTRIBUTE_ALIGN(32);
+	s32 ret = SHA_InitializeContext(&context);
+	if(ret < 0)
+		return ret;
+
+	return SHA_Finalize(&context, data, data_size, message_digest);
```

Comment:
this needs to handle large input data by looping over it with SHA_Input and then a final SHA_Finalize

Reply:
it already does, check the code of `SHA_ExecuteCommand`

Reply:
oh i see, when you said “hence the splitting up of commands”, you meant for the ipc rather than for libogc’s api, where the splitting isn’t necessary. i thought you meant “that’s why the api has those commands split up”